[WIP] JAX implementation

bamos opened the pull request:
Hey! Hope you all have been doing well. I've been using JAX and was finally able to connect this in for #7. This initial version is pretty similar to the PyTorch one and was easy with some of the recent JAX developments on enabling custom JAX primitives with a non-traced vjp. The interface follows our PyTorch/TF layers has an outer function that takes the CVXPY objects and then returns a callable into a primitive for the problem.

Let me know if you have any thoughts on this initial version! There are a few things left I'll finish up over the next few days:

+ [x] Add DGP
+ [x] Add the rest of the PyTorch tests and add JAX to the CI
+ [x] Update the README (image and JAX example)
+ [x] Document current limitations somewhere (e.g. can't use this with the JIT/vmap, and can't do higher-order derivatives)
+ [x] Add a sample notebook

Some useful JAX references:
+ https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html
+ https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html
+ https://github.com/google/jax/issues/1142 and https://github.com/google/jax/issues/3415.  @mattjj and @shoyer, your responses in these threads were extremely helpful! Let me know if there's anything else I should watch out for in this PR
+ https://gist.github.com/shoyer/5f72853c2788e99e785f4737ee8a6ae1 -- JAX wrapper around TF2

\cc @bodono

akshayka commented:
This looks great!! Thanks so much for taking this on. I'd love to have this merged in, once it's finished.

sbarratt commented:
> This looks great!! Thanks so much for taking this on. I'd love to have this merged in, once it's finished.

+1

shoyer commented:
This is great!

A few ideas for extending this:
- If you want to add support for `jit`, I would suggest looking into jax.experimental.host_callback: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html
- You might also write the forward operator as a JAX primitive, which would allow you do define a batching rule to support `vmap`.

bamos commented:
@akshayka @sbarratt, can you do a review through this now? The main layer and all tests are implemented and passing (with GP support), and I'll add the last piece with a sample/tutorial notebook tomorrow.

akshayka commented on README.md:
```diff
@@ -5,7 +5,7 @@
 # cvxpylayers
```
At some point we should update the CVXPY Layers logo to include JAX :)

bamos replied:
Done :)

akshayka commented on cvxpylayers/jax/cvxpylayer.py:
```diff
@@ -0,0 +1,322 @@
+            raise ValueError('Problem must be DPP.')
+
+    if not set(problem.parameters()) == set(parameters):
+        raise ValueError("The layer's parameters must exactly match "
+                         "problem.parameters")
+    if not set(variables).issubset(set(problem.variables())):
+        raise ValueError("Argument variables must be a subset of "
+                         "problem.variables")
+    if not isinstance(parameters, list) and \
+            not isinstance(parameters, tuple):
+        raise ValueError("The layer's parameters must be provided as "
+                         "a list or tuple")
+    if not isinstance(variables, list) and \
+            not isinstance(variables, tuple):
+        raise ValueError("The layer's variables must be provided as "
+                         "a list or tuple")
+
+    var_dict = {v.id for v in variables}
+
+    # Construct compiler
+    param_order = parameters
+    if gp:
+        for param in parameters:
+            if param.value is None:
+                raise ValueError("An initial value for each parameter is "
+                                 "required when gp=True.")
+        data, solving_chain, _ = problem.get_problem_data(
+            solver=cp.SCS, gp=True)
+        compiler = data[cp.settings.PARAM_PROB]
+        dgp2dcp = solving_chain.get(cp.reductions.Dgp2Dcp)
+        param_ids = [p.id for p in compiler.parameters]
+        old_params_to_new_params = (
+            dgp2dcp.canon_methods._parameters
+        )
+    else:
+        data, _, _ = problem.get_problem_data(solver=cp.SCS)
+        compiler = data[cp.settings.PARAM_PROB]
+        param_ids = [p.id for p in param_order]
+        dgp2dcp = None
+    cone_dims = dims_to_solver_dict(data["dims"])
```
At some point we should abstract this into a common function, used by all three layers. Maintaining three separate versions of the same code won't be fun.

No pressure to do this now (though you're more than welcome to do so, if you want to).

akshayka reviewed (approved):
LGTM! It's awesome that we'll finally have a JAX layer.

With this change we now have a lot of duplicated code, across the three layers. It would be great if this code could be unified and re-used by the three layers. You don't need to do that for this PR, though you're more than welcome to do so if you have the inclination.

sbarratt reviewed (approved):
Looks good to me!

bamos commented:
I just pushed a notebook with an intro to the JAX version, squashed the commits, and added followup development issues. I may put in a few last tweaks and will merge in soon once CI passes! Let me know if you all have any other comments or anything else you want to get in before this is merged.

froystig commented:
To add to @shoyer's comment:
> * If you want to add support for `jit`, I would suggest looking into jax.experimental.host_callback: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html

Another approach is to use XLA's custom call via jax. [This tutorial](https://github.com/dfm/extending-jax) covers the topic well.

akshayka commented:
> I just pushed a notebook with an intro to the JAX version, squashed the commits, and added followup development issues. I may put in a few last tweaks and will merge in soon once CI passes! Let me know if you all have any other comments or anything else you want to get in before this is merged.

Awesome! I'll take a look tonight or tomorrow morning. Excited for this to be merged in.
